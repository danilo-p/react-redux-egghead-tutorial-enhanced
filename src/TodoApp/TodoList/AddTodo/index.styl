.AddTodo__Input
    margin-right: 7.5px;
