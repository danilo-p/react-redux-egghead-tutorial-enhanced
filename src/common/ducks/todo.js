// Initial State
export const INITIAL_STATE = {};

// Actions
export const ADD = 'ADD_TODO';
export const TOGGLE = 'TOGGLE_TODO';

// Reducer
export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
  case ADD:
    return {
      id: action.id,
      text: action.text,
      completed: false,
    };
  case TOGGLE:
    if (state.id !== action.id) {
      return state;
    }
    return {
      ...state,
      completed: !state.completed,
    };
  default:
    return state;
  }
};

export const createTodoId = text => Date.now() + text;

// Action Creators
export function addTodo(text, id = createTodoId(text)) {
  return {
    type: ADD,
    id,
    text,
  };
}

export const toggleTodo = id => ({
  type: TOGGLE,
  id,
});
