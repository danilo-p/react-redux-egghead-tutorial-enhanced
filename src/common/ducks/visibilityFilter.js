/**
 * The "visibilityFilter" duck.
 *
 * @module common/ducks/visibilityFilter
 */

/* ======================================
=            Initial State              =
====================================== */

/** "visibilityFilter" duck initial state */
export const INITIAL_STATE = 'SHOW_ALL';

/* ======  End of Initial State  ====== */


/* =====================================
=            Action Types              =
===================================== */

/** SET_VISIBILITY_FILTER action type */
export const SET = 'SET_VISIBILITY_FILTER';

/* ======  End of Action Types  ====== */


/* ================================
=            Reducer              =
================================ */

/**
 * The "visibilityFilter" duck reducer
 *
 * @memberof module:common/ducks/visibilityFilter
 * @param {Object} [state=INITIAL_STATE] The current state
 * @param {Object} action The action to be handled
 * @returns {Object} The new state
 */
export default function reducer(state = INITIAL_STATE, action) {
  switch (action.type) {
  case SET:
    return action.filter;
  default:
    return state;
  }
}

/* ======  End of Reducer  ====== */


/* ========================================
=            Action Creators              =
======================================== */

/**
 * Creates a new SET_VISIBILITY_FILTER action.
 *
 * @memberof module:common/ducks/visibilityFilter
 * @param {String} filter The visibility filter
 * @returns {Object} The new SET_VISIBILITY_FILTER action
 */
export const setVisibilityFilter = filter => ({
  type: 'SET_VISIBILITY_FILTER',
  filter,
});

/* ======  End of Action Creators  ====== */
