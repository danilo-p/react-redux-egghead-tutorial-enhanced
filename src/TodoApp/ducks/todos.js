import todo, {
  ADD,
  TOGGLE,
} from './todo';

// Actions
export const REMOVE = 'REMOVE_TODO';
export { ADD, TOGGLE };

// Reducer
export default (
  state = [
    {
      id: 0,
      text: 'test0',
      completed: false,
    },
    {
      id: 1,
      text: 'test1',
      completed: false,
    },
    {
      id: 2,
      text: 'test2',
      completed: false,
    },
  ],
  action,
) => {
  switch (action.type) {
  case ADD:
    return [
      ...state,
      todo(undefined, action),
    ];
  case TOGGLE:
    return state.map(t => todo(t, action));
  case REMOVE:
    return []; // TODO: implement the remove todo
  default:
    return state;
  }
};
