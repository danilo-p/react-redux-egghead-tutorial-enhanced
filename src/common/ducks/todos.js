import todo, {
  ADD,
  TOGGLE,
} from './todo';

// Actions
export const REMOVE = 'REMOVE_TODO';
export { ADD, TOGGLE };

// Reducer
export default (
  state = [],
  action,
) => {
  switch (action.type) {
  case ADD:
    return [
      ...state,
      todo(undefined, action),
    ];
  case TOGGLE:
    return state.map(t => todo(t, action));
  case REMOVE:
    return []; // TODO: implement the remove todo
  default:
    return state;
  }
};
