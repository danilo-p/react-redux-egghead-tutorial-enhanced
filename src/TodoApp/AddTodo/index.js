import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Input, Button } from 'semantic-ui-react';
import { addTodo } from '../ducks/todo';
import styles from './index.styl';

class AddTodo extends React.Component {
  constructor() {
    super();
    this.state = {
      inputText: '',
    };
    this.onInputChange = this.onInputChange.bind(this);
    this.onButtonClick = this.onButtonClick.bind(this);
  }

  onInputChange(event) {
    this.setState({
      inputText: event.target.value,
    });
  }

  onButtonClick() {
    this.props.dispatch(addTodo(this.state.inputText));
    this.setState({
      inputText: '',
    });
  }

  render() {
    return (
      <div className="AddTodo">
        <Input
          placeholder="Remind to feed the..."
          className={styles.AddTodo__Input}
          value={this.state.inputText}
          onChange={this.onInputChange}
          onKeyPress={(event) => {
            if (event.key === 'Enter') {
              this.onButtonClick();
            }
          }}
        />
        <Button
          primary
          onClick={this.onButtonClick}
          className="AddTodo__Button"
        >
          Add Todo
        </Button>
      </div>
    );
  }
}

AddTodo.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

export default connect()(AddTodo);
