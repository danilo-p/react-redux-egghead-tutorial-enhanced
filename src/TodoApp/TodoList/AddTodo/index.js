import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Input, Button } from 'semantic-ui-react';
import { addTodo } from 'common/ducks/todos';
import styles from './index.styl';

/**
 * Used for add a new todo with an input.
 *
 * @class AddTodo
 * @extends {React.Component}
 *
*/
class AddTodo extends React.Component {
  /**
   * Creates an instance of AddTodo.
   * @memberof AddTodo
   */
  constructor() {
    super();
    this.state = {
      inputText: '',
    };
    this.onInputChange = this.onInputChange.bind(this);
    this.onButtonClick = this.onButtonClick.bind(this);
  }

  /**
   * Handles the input change and updates the component state.
   *
   * @param {any} event The change event containing the current value of the
   * input.
   * @memberof AddTodo
   * @return {void}
   */
  onInputChange(event) {
    this.setState({
      inputText: event.target.value,
    });
  }
  /**
   * Handles the button click and dispatch an addTodo action if allowed.
   *
   * @memberof AddTodo
   * @return {void}
   */
  onButtonClick() {
    const { inputText } = this.state;
    if (inputText) {
      this.props.dispatch(addTodo(inputText));
      this.setState({
        inputText: '',
      });
    }
  }
  /**
   * Renders the component
   *
   * @memberof AddTodo
   * @return {jsx} The component template
   */
  render() {
    return (
      <div className="AddTodo">
        <Input
          action
          placeholder="Remind to feed the..."
          className={styles.AddTodo__Input}
          value={this.state.inputText}
          onChange={this.onInputChange}
          onKeyPress={(event) => {
            if (event.key === 'Enter') {
              this.onButtonClick();
            }
          }}
        >
          <input />
          <Button
            primary
            onClick={this.onButtonClick}
            className="AddTodo__Button"
          >
            Add Todo
          </Button>
        </Input>
      </div>
    );
  }
}

AddTodo.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

export default connect()(AddTodo);
