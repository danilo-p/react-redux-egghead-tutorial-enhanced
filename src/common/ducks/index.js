/**
 * Module that bundles all the ducks in one reducer. It imports all the ducks
 * and combine their reducers with the combineReducers method of the redux
 * library.
 *
 * @module common/ducks
 */

import { combineReducers } from 'redux';

import todo from './todo';
import todos from './todos';
import visibilityFilter from './visibilityFilter';

// Here all the reducers are combined into a single one reducer that is
// exported as default and used to create the application store.
export default combineReducers({
  todo,
  todos,
  visibilityFilter,
});
