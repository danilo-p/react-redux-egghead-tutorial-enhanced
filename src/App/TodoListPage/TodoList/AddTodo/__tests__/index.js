import React from 'react';
import { Input, Button } from 'semantic-ui-react';
import { INITIAL_STATE } from 'common/ducks';
import { addTodo as addTodoActionCreator } from 'common/ducks/todos';
import configureMockStore from 'redux-mock-store';
import AddTodo from '../';

const mockStore = configureMockStore();

describe('AddTodo', () => {
  let addTodo = null;
  let store = null;

  beforeEach(() => {
    store = mockStore(INITIAL_STATE);
    addTodo = mount(<AddTodo store={store} />);
  });

  it('should render the component', () => {
    expect(addTodo).toMatchSnapshot();
  });

  it('always renders a div that contains everything else', () => {
    const divs = addTodo.find('div');
    expect(divs.length).toBeGreaterThan(0);
  });

  it('always renders a "Input"', () => {
    expect(addTodo.find(Input).length).toBe(1);
  });

  it('always renders a "Button"', () => {
    expect(addTodo.find(Button).length).toBe(1);
  });

  describe('when the user is typing', () => {
    const changeInputValue = (value) => {
      const input = addTodo.find(Input).first();
      const eventMock = { target: { value } };
      input.find('input').first().simulate('change', eventMock);
    };

    const hitKey = (key) => {
      const input = addTodo.find(Input).first();
      const eventMock = { key };
      input.find('input').first().simulate('keyPress', eventMock);
    };

    it('should update the input value', () => {
      const value = 'Lorem ipsum dolor';
      changeInputValue(value);
      setTimeout(() => { // Wait for state update
        expect(addTodo.state().inputText).toBe(value);
      }, 1000);
    });

    describe('and hits any other key that is not "Enter"', () => {
      it('should update the input value with the typed key', () => {
        const key = 'A';
        hitKey(key);
        setTimeout(() => { // Wait for state update
          expect(addTodo.state().inputText).toBe(key);
        }, 1000);
      });
    });

    describe('and tries to add a new todo', () => {
      const value = 'Lorem ipsum dolor';

      describe('hiting the "Enter" key', () => {
        describe('and the input is empty', () => {
          it('should not dispatch any action', () => {
            hitKey('Enter');
            setTimeout(() => { // Wait for state update
              expect(store.getActions().length).toBe(0);
            }, 1000);
          });
        });

        describe('and the input is not empty', () => {
          beforeEach(() => {
            changeInputValue(value);
            hitKey('Enter');
          });

          it('should dispatch a new ADD_TODO action', () => {
            setTimeout(() => { // Wait for state update
              const actions = store.getActions();
              expect(actions.length).toBe(1);

              const addTodoAction = actions[0];
              expect(addTodoAction)
                .toEqual(addTodoActionCreator(value, addTodoAction.id));
            }, 1000);
          });

          it('should clear the state value', () => {
            setTimeout(() => { // Wait for state update
              expect(addTodo.state().inputText).toBe('');
            }, 1000);
          });
        });
      });

      describe('clicking on the "Add Todo" button', () => {
        const clickAddTodoButton = () => {
          const button = addTodo.find(Button).first();
          button.simulate('click');
        };

        describe('and the input is empty', () => {
          it('should not dispatch any action', () => {
            clickAddTodoButton();
            setTimeout(() => { // Wait for state update
              expect(store.getActions().length).toBe(0);
            }, 1000);
          });
        });

        describe('and the input is not empty', () => {
          beforeEach(() => {
            changeInputValue(value);
            clickAddTodoButton();
          });

          it('should dispatch a new ADD_TODO action', () => {
            setTimeout(() => { // Wait for state update
              const actions = store.getActions();
              expect(actions.length).toBe(1);

              const addTodoAction = actions[0];
              expect(addTodoAction)
                .toEqual(addTodoActionCreator(value, addTodoAction.id));
            }, 1000);
          });

          it('should clear the state value', () => {
            setTimeout(() => { // Wait for state update
              expect(addTodo.state().inputText).toBe('');
            }, 1000);
          });
        });
      });
    });
  });
});
