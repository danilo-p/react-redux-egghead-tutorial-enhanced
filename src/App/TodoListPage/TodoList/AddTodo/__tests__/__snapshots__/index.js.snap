// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo should shallow render the component 1`] = `
<Connect(AddTodo)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AddTodo
    dispatch={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <div
      className="AddTodo"
    >
      <Input
        action={
          <Button
            as="button"
            className="AddTodo__Button"
            onClick={[Function]}
            primary={true}
          >
            Add Todo
          </Button>
        }
        className="AddTodo__Input"
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Remind to feed the..."
        type="text"
        value=""
      >
        <div
          className="ui action input AddTodo__Input"
        >
          <input
            disabled={undefined}
            key="text"
            onChange={[Function]}
            onKeyPress={[Function]}
            placeholder="Remind to feed the..."
            tabIndex={undefined}
            type="text"
            value=""
          />
          <Button
            as="button"
            className="AddTodo__Button"
            onClick={[Function]}
            primary={true}
          >
            <button
              className="ui primary button AddTodo__Button"
              disabled={undefined}
              onClick={[Function]}
              role="button"
              tabIndex={undefined}
            >
              Add Todo
            </button>
          </Button>
        </div>
      </Input>
    </div>
  </AddTodo>
</Connect(AddTodo)>
`;
